File: record-api/src/Infrastructure/Data/Repositories/RecordRepository.cs
using Microsoft.AspNetCore.Http;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Dapper;
using Contracts;
using System.Data;
using System.IO;
using Application.Data;
using Application.Data.Repositories;
using Contracts.Request;
using System.Text.Json;
using Domain.ValueObjects;
using Domain.Entities;
using Application.Data.DataTransfer;
using Application.Data.DataTransfer.Record;

namespace Infrastructure.Data.Repositories
{
    internal class RecordRepository : IRecordRepository
    {
        private readonly IConnectionFactory _connectionFactory;

        public RecordRepository(IConnectionFactory connectionFactory)
        {
            _connectionFactory = connectionFactory;
        }

        public async Task<PaginationEnvelope<RecordGetAll>>
            GetWithRecordTypeIdAsync(Pagination pagination, int recordTypeId)
        {
            using var connection = _connectionFactory.NewConnection;

            (int skip, int take) = pagination.GetSkipTake();

            var parameter = new DynamicParameters();
            parameter.Add("param_skip", skip);
            parameter.Add("param_take", take);
            parameter.Add("param_recordTypeId", recordTypeId);
            parameter.Add("param_total", dbType: DbType.Int32, direction: ParameterDirection.Output);

            var result = await connection
                .QueryAsync<string>(StoredProcedures.Record.GetWithRecordTypeId,
                    param: parameter,
                    commandType: CommandType.StoredProcedure);

            int total = parameter.Get<int>("param_total");

            var records = JsonSerializer.Deserialize<IEnumerable<RecordGetAll>>(string.Join("", result),
                new(JsonSerializerDefaults.Web));

            return new PaginationEnvelope<RecordGetAll>
            {
                Data = records,
                PageNumber = pagination.PageNumber,
                PageSize = pagination.PageSize,
                Total = total
            };
        }

        public async Task<RecordGetLoaded> GetLoadedAsync()
        {
            using var connection = _connectionFactory.NewConnection;

            var result = await connection.QueryAsync<string>(StoredProcedures.Record.GetLoaded,
                commandType: CommandType.StoredProcedure);

            if (!result.Any())
            {
                return null;
            }
            
            var loadedRecord = JsonSerializer
                .Deserialize<RecordGetLoaded>(string.Join("", result),
                    new(JsonSerializerDefaults.Web));

            return loadedRecord;
        }

        public async Task<RecordGetById> GetByIdAsync(int id)
        {
            using var connection = _connectionFactory.NewConnection;

            var result = await connection.QueryAsync<string>(StoredProcedures.Record.GetById,
                param: new { param_id = id },
                commandType: CommandType.StoredProcedure);

            if (!result.Any())
            {
                return null;
            }

            var record = JsonSerializer.Deserialize<RecordGetById>(string.Join("", result),
                new(JsonSerializerDefaults.Web));

            return record;
        }

        public async Task<byte[]> DownloadAsync(int id)
        {
            using var connection = _connectionFactory.NewConnection;

            var zipContent = await connection.QuerySingleOrDefaultAsync<byte[]>(StoredProcedures.Record.Download,
                param: new { param_id = id },
                commandType: CommandType.StoredProcedure);

            return zipContent;
        }

        public async Task<int> CreateAsync(RecordCreate createData)
        {
            using var connection = _connectionFactory.NewConnection;

            var parameter = new DynamicParameters();
            parameter.Add("param_modelFileName", createData.ModelFileName);
            parameter.Add("param_LabelFileName", createData.LabelFileName);
            parameter.Add("param_zipContent", createData.ZipContent);
            parameter.Add("param_recordTypeId", createData.RecordTypeId);

            return await connection.QuerySingleAsync<int>(StoredProcedures.Record.Create,
                param: parameter,
                commandType: CommandType.StoredProcedure);
        }

        public async Task<int> SetLoadedAsync(int recordId)
        {
            using var connection = _connectionFactory.NewConnection;

            return await connection.ExecuteAsync(StoredProcedures.Record.SetLoaded,
                param: new { param_recordId = recordId },
                commandType: CommandType.StoredProcedure);
        }

        public async Task<int> DeleteAsync(int id)
        {
            using var connection = _connectionFactory.NewConnection;

            return await connection.ExecuteAsync(StoredProcedures.Record.Delete,
                param: new { param_id = id },
                commandType: CommandType.StoredProcedure);
        }
    }
}

File: vue-app/src/api/repositories/recordRepository.ts
import { useDownload, useFetch } from '../../composable'
import { PaginationEnvelope, Id, PaginationRequest } from './common'

export type Record = {
  id: number
  modelFileName: string
  labelFileName: string
  loaded: boolean
  recordTypeId: number
  recordType: string
}

export const recordRepository = {
  getWidthRecordTypeId<T extends boolean>(
    immediate: T,
    recordTypeId: Id,
    pagination: PaginationRequest = { pageNumber: 1, pageSize: 100 }
  ) {
    return useFetch<PaginationEnvelope<Record>>(
      `/record/type/${recordTypeId}`,
      pagination
    )
      .get()
      .json(immediate)
  },
  getLoaded<T extends boolean>(immediate: T) {
    return useFetch<Record>('/record/loaded').get().json(immediate)
  },
  getById<T extends boolean>(immediate: T, id: Id) {
    return useFetch<Record>(`/record/${id}`).get().json(immediate)
  },
  async download(id: Id) {
    const { data } = await useFetch<Blob>(`/record/download/${id}`)
      .get()
      .blob(false).promise

    if (data) {
      useDownload(data, 'model.zip')
    }
  },
  upload<T extends boolean>(
    immediate: T,
    modelTypeId: Id,
    model: File,
    label: File
  ) {
    const formData = new FormData()
    formData.append('modelTypeId', modelTypeId.toString())
    formData.append('model', model)
    formData.append('label', label)

    return useFetch<void>('/record').post({ body: formData }).text(immediate)
  },
  setLoaded<T extends boolean>(immediate: T, id: Id) {
    return useFetch<void>(`/record/${id}`).put().text(immediate)
  },
  delete<T extends boolean>(immediate: T, id: Id) {
    return useFetch<void>(`/record/${id}`).delete().text(immediate)
  }
}
